scripts/west_commands/runners/nrfjprog.py

```diff
@@ -150,68 +163,131 @@ def get_board_snr(self, glob):
-                    commands.append(program_cmd + ['--sectorerase'])
+                erase_arg = '--sectorerase'
 
+        if self.family == 'NRF53':
+            if self.build_conf.get('CONFIG_SOC_NRF5340_CPUAPP', False):
+                coprocessor = 'CP_APPLICATION'
+            elif self.build_conf.get('CONFIG_SOC_NRF5340_CPUNET', False):
+                coprocessor = 'CP_NETWORK'
+            else:
+                # When it's time to update this file, it would probably be best
+                # to handle this by adding common 'SOC_NRF53X_CPUAPP'
+                # and 'SOC_NRF53X_CPUNET' options, so we don't have to
+                # maintain a list of SoCs in this file too.
+                raise RuntimeError(f'unknown nRF53; update {__file__}')
+            coprocessor_args = ['--coprocessor', coprocessor]
+        else:
+            coprocessor_args = []
+
+        # It's important for tool_opt to come last, so it can override
+        # any options that we set here.
+        try:
+            self.check_call(['nrfjprog', '--program', self.hex_, erase_arg,
+                             '-f', self.family, '--snr', self.snr] +
+                            coprocessor_args + self.tool_opt)
+        except subprocess.CalledProcessError as cpe:
+            if cpe.returncode == UnavailableOperationBecauseProtectionError:
+                self.logger.error('Flashing failed because the target '
+                                  'must be recovered. Please run '
+                                  '"west flash --recover" instead.')
+            raise
+
+    def reset_target(self):
         if self.family == 'NRF52' and not self.softreset:
-            commands.extend([
-                # Enable pin reset
-                ['nrfjprog', '--pinresetenable', '-f', self.family,
-                 '--snr', board_snr],
-            ])
+            self.check_call(['nrfjprog', '--pinresetenable', '-f', self.family,
+                             '--snr', self.snr])  # Enable pin reset
 
         if self.softreset:
-            commands.append(['nrfjprog', '--reset', '-f', self.family,
-                             '--snr', board_snr])
+            self.check_call(['nrfjprog', '--reset', '-f', self.family,
+                             '--snr', self.snr])
         else:
-            commands.append(['nrfjprog', '--pinreset', '-f', self.family,
-                             '--snr', board_snr])
+            self.check_call(['nrfjprog', '--pinreset', '-f', self.family,
+                             '--snr', self.snr])
+
+    def do_run(self, command, **kwargs):
+        self.require('nrfjprog')
+        self.build_conf = BuildConfiguration(self.cfg.build_dir)
+        if not os.path.isfile(self.hex_):
+            raise RuntimeError(
+                f'Cannot flash; hex file ({self.hex_}) does not exist. '
+                'Try enabling CONFIG_BUILD_OUTPUT_HEX.')
+
+        self.ensure_snr()
+        self.ensure_family()
+        self.check_force_uicr()
```

Comment:
Checking the uicr here actually mean that user might be printed with:
```
The hex file contains data placed in the UICR, which
needs a full erase before reprogramming. Run west
flash again with --force or --erase.
```
but that could end up not being sufficient if readback protection in active, thus resulting in this message if user add `--erase` or `--force`:
```
Flashing failed because the target must be recovered. Please run "west flash --recover" instead.
```

Before informing the user to do `--erase` or `--force`, we could check if readback protect in active, and then tell him to recover instead.

A simple:
```
nrfjprog --memrd 0x0 --n 4
```
will actually tell you if readback protection is enabled (maybe @simtind has a more clever way to check readback protection)

Reply:
Not using nrfjprog.exe, unfortunately, trying and failing is the only way.

Reply:
> Not using nrfjprog.exe, unfortunately, trying and failing is the only way.

This combined with @tejlmand's request seems to point at running the `--memrd` unconditionally and erroring out if `--recover` isn't given. If the `--memrd` succeeds, then we error out unless `--erase` or `--force`.

Right?

Reply:
yep, that was my thinking.